File: src/main/java/org/meetkt/catalogue/domain/model/Catalogue.java
package org.meetkt.catalogue.domain.model;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;

import static java.util.stream.Collectors.toMap;

public class Catalogue {

    private final Map<String, Product> productByBarcode;

    public Catalogue(List<Product> products) {
        this.productByBarcode =
                Objects.requireNonNull(products).stream().collect(toMap(Product::barcode, Function.identity()));
    }

    public Optional<Product> productWith(String barcode) {
        return Optional.ofNullable(productByBarcode.get(barcode));
    }
}

File: src/main/kotlin/in/meetkt/catalogue/domain/model/Catalogue.kt
package `in`.meetkt.catalogue.domain.model

import arrow.core.Option

class Catalogue(products: List<Product>) {

    private val productByBarcode: Map<String, Product> = products.map { it.barcode to it }.toMap()

    fun productWith(barcode: String): Option<Product> = Option.fromNullable(productByBarcode[barcode])
}
